File: packages/sdk/go/coordinator.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Coordinator REST client for the Go SDK.

package sdk

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// CoordinatorClient is the Caracal coordinator REST client.
type CoordinatorClient struct {
	BaseURL    string
	HTTPClient *http.Client
}

func (c *CoordinatorClient) http() *http.Client {
	if c.HTTPClient != nil {
		return c.HTTPClient
	}
	return http.DefaultClient
}

// AgentKind distinguishes the agent session lifecycle.
type AgentKind string

const (
	KindService   AgentKind = "service"
	KindInstance  AgentKind = "instance"
	KindEphemeral AgentKind = "ephemeral"
)

// SpawnRequest parameters for coordinator agent spawn.
type SpawnRequest struct {
	ZoneID           string
	ApplicationID    string
	SubjectSessionID string
	ParentID         string
	Kind             AgentKind
	TTLSeconds       int
	Metadata         map[string]any
	Labels           []string
	IdempotencyKey   string
}

// SpawnResponse from the coordinator.
type SpawnResponse struct {
	AgentSessionID string `json:"agent_session_id"`
}

// DelegationConstraints narrows a delegation edge.
type DelegationConstraints struct {
	Resources      []string
	MaxDepth       int
	MaxHops        int
	TTLSeconds     int
	Budget         int
	PolicyApproved bool
	ExpiresAt      string
	BroadReason    string
}

func (d *DelegationConstraints) toWire() map[string]any {
	out := map[string]any{}
	if d.Resources != nil {
		out["resources"] = d.Resources
	}
	if d.MaxDepth > 0 {
		out["max_depth"] = d.MaxDepth
	}
	if d.MaxHops > 0 {
		out["max_hops"] = d.MaxHops
	}
	if d.TTLSeconds > 0 {
		out["ttl_seconds"] = d.TTLSeconds
	}
	if d.Budget > 0 {
		out["budget"] = d.Budget
	}
	if d.PolicyApproved {
		out["policy_approved"] = d.PolicyApproved
	}
	if d.ExpiresAt != "" {
		out["expires_at"] = d.ExpiresAt
	}
	if d.BroadReason != "" {
		out["broad_reason"] = d.BroadReason
	}
	return out
}

// DelegationRequest parameters for coordinator delegation edge creation.
type DelegationRequest struct {
	ZoneID                string
	IssuerApplicationID   string
	SourceSessionID       string
	TargetSessionID       string
	ReceiverApplicationID string
	ParentEdgeID          string
	ResourceID            string
	Scopes                []string
	Constraints           *DelegationConstraints
	TTLSeconds            int
}

// DelegationResponse from the coordinator.
type DelegationResponse struct {
	DelegationEdgeID string `json:"delegation_edge_id"`
}

// SpawnAgent calls POST /zones/:zoneId/agents.
func SpawnAgent(ctx context.Context, client *CoordinatorClient, bearer string, req SpawnRequest) (SpawnResponse, error) {
	body := map[string]any{
		"application_id": req.ApplicationID,
	}
	if req.Kind != "" {
		body["kind"] = string(req.Kind)
	}
	if req.SubjectSessionID != "" {
		body["subject_session_id"] = req.SubjectSessionID
	}
	if req.ParentID != "" {
		body["parent_id"] = req.ParentID
	}
	if req.TTLSeconds > 0 {
		body["ttl_seconds"] = req.TTLSeconds
	}
	if req.Metadata != nil {
		body["metadata"] = req.Metadata
	}
	if len(req.Labels) > 0 {
		body["labels"] = req.Labels
	}

	extra := map[string]string{}
	key := req.IdempotencyKey
	if key == "" {
		key = deriveIdempotencyKey(req)
	}
	if key != "" {
		extra["Idempotency-Key"] = key
	}

	var out SpawnResponse
	err := doJSON(ctx, client, "POST", fmt.Sprintf("/zones/%s/agents", req.ZoneID), bearer, body, extra, &out)
	return out, err
}

// deriveIdempotencyKey produces a stable key for SDK-issued spawn retries.
// Returns empty when no stable inputs are present: in that case the caller's
// retry would still require a fresh session.
func deriveIdempotencyKey(req SpawnRequest) string {
	if req.SubjectSessionID == "" && req.ParentID == "" {
		return ""
	}
	seed := req.ApplicationID + "|" + req.SubjectSessionID + "|" + req.ParentID + "|" + string(req.Kind) + "|" + strings.Join(req.Labels, ",")
	sum := sha256.Sum256([]byte(seed))
	return hex.EncodeToString(sum[:])
}

// TerminateAgent calls DELETE /zones/:zoneId/agents/:id.
func TerminateAgent(ctx context.Context, client *CoordinatorClient, bearer, zoneID, agentSessionID string) error {
	return doJSON(ctx, client, "DELETE", fmt.Sprintf("/zones/%s/agents/%s", zoneID, agentSessionID), bearer, nil, nil, nil)
}

// HeartbeatAgent renews a service agent's lease. A service session is reaped by
// the coordinator if it stops heartbeating before the lease expires.
func HeartbeatAgent(ctx context.Context, client *CoordinatorClient, bearer, zoneID, agentSessionID string) error {
	body := map[string]any{"status": "healthy"}
	return doJSON(ctx, client, "POST", fmt.Sprintf("/zones/%s/agents/%s/heartbeat", zoneID, agentSessionID), bearer, body, nil, nil)
}

// CreateDelegation calls POST /zones/:zoneId/delegations.
func CreateDelegation(ctx context.Context, client *CoordinatorClient, bearer string, req DelegationRequest) (DelegationResponse, error) {
	body := map[string]any{
		"issuer_application_id":   req.IssuerApplicationID,
		"source_session_id":       req.SourceSessionID,
		"target_session_id":       req.TargetSessionID,
		"receiver_application_id": req.ReceiverApplicationID,
		"scopes":                  req.Scopes,
	}
	if req.ResourceID != "" {
		body["resource_id"] = req.ResourceID
	}
	if req.ParentEdgeID != "" {
		body["parent_edge_id"] = req.ParentEdgeID
	}
	if req.Constraints != nil {
		body["constraints"] = req.Constraints.toWire()
	}
	if req.TTLSeconds > 0 {
		body["ttl_seconds"] = req.TTLSeconds
	}

	var out DelegationResponse
	err := doJSON(ctx, client, "POST", fmt.Sprintf("/zones/%s/delegations", req.ZoneID), bearer, body, nil, &out)
	return out, err
}

func doJSON(ctx context.Context, client *CoordinatorClient, method, path, bearer string, body any, extraHeaders map[string]string, out any) error {
	var bodyReader io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return err
		}
		bodyReader = bytes.NewReader(b)
	}

	req, err := http.NewRequestWithContext(ctx, method, client.BaseURL+path, bodyReader)
	if err != nil {
		return err
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	req.Header.Set("Authorization", "Bearer "+bearer)
	for k, v := range extraHeaders {
		req.Header.Set(k, v)
	}

	resp, err := client.http().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		raw, readErr := io.ReadAll(resp.Body)
		if readErr != nil {
			return fmt.Errorf("coordinator %s %s: %d (reading response body: %w)", method, path, resp.StatusCode, readErr)
		}
		return fmt.Errorf("coordinator %s %s: %d %s", method, path, resp.StatusCode, raw)
	}

	if out != nil {
		return json.NewDecoder(resp.Body).Decode(out)
	}
	return nil
}

File: packages/sdk/ts/src/coordinator.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * Coordinator REST client used by SDK primitives.
 */

import { createHash } from 'node:crypto'
import type { JsonObject } from './json.js'

export interface CoordinatorClient {
  baseUrl: string
  fetchImpl?: typeof fetch
}

export const Lifecycle = {
  Task: 'task',
  Service: 'service',
} as const

export type Lifecycle = (typeof Lifecycle)[keyof typeof Lifecycle]

export interface DelegationConstraints {
  resources?: string[]
  maxDepth?: number
  maxHops?: number
  ttlSeconds?: number
  budget?: number
  policyApproved?: boolean
  expiresAt?: string
  broadReason?: string
}

export interface SpawnRequest {
  zoneId: string
  applicationId: string
  subjectSessionId?: string
  parentId?: string
  lifecycle?: Lifecycle
  ttlSeconds?: number
  metadata?: JsonObject
  labels?: string[]
  idempotencyKey?: string
  inheritParentEdgeId?: string
}

export interface SpawnResponse {
  agent_session_id: string
  delegation_edge_id?: string | null
}

export interface DelegationRequest {
  zoneId: string
  issuerApplicationId: string
  sourceSessionId: string
  targetSessionId: string
  receiverApplicationId: string
  parentEdgeId?: string
  resourceId?: string
  scopes: string[]
  constraints?: DelegationConstraints
  ttlSeconds?: number
}

export interface DelegationResponse {
  delegation_edge_id: string
}

async function call<T>(
  client: CoordinatorClient,
  method: string,
  path: string,
  bearer: string,
  body?: unknown,
  extraHeaders?: Record<string, string>,
): Promise<T> {
  const fetchFn = client.fetchImpl ?? fetch
  const headers: Record<string, string> = {
    'content-type': 'application/json',
    authorization: `Bearer ${bearer}`,
    ...(extraHeaders ?? {}),
  }
  const res = await fetchFn(`${client.baseUrl}${path}`, {
    method,
    headers,
    body: body ? JSON.stringify(body) : undefined,
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`coordinator ${method} ${path} failed: ${res.status} ${text}`)
  }
  return (await res.json()) as T
}

export async function spawnAgent(client: CoordinatorClient, bearer: string, req: SpawnRequest): Promise<SpawnResponse> {
  const key = req.idempotencyKey ?? deriveIdempotencyKey(req)
  const headers = key ? { 'idempotency-key': key } : undefined
  const res = await call<SpawnResponse>(
    client,
    'POST',
    `/zones/${encodeURIComponent(req.zoneId)}/agents`,
    bearer,
    {
      application_id: req.applicationId,
      subject_session_id: req.subjectSessionId,
      parent_id: req.parentId,
      lifecycle: req.lifecycle,
      ttl_seconds: req.ttlSeconds,
      metadata: req.metadata,
      labels: req.labels,
      inherit_parent_edge_id: req.inheritParentEdgeId,
    },
    headers,
  )
  return res
}

/**
 * Stable key for SDK-issued spawn retries. Returns undefined when no stable
 * inputs are present: in that case the caller's retry would still require a
 * fresh session.
 */
function deriveIdempotencyKey(req: SpawnRequest): string | undefined {
  if (!req.subjectSessionId && !req.parentId) return undefined
  const seed = [
    req.applicationId,
    req.subjectSessionId ?? '',
    req.parentId ?? '',
    String(req.lifecycle ?? ''),
    (req.labels ?? []).join(','),
  ].join('|')
  return sha256Hex(seed)
}

function sha256Hex(input: string): string {
  return createHash('sha256').update(input).digest('hex')
}

export async function terminateAgent(client: CoordinatorClient, bearer: string, zoneId: string, agentSessionId: string): Promise<void> {
  const fetchFn = client.fetchImpl ?? fetch
  const del = await fetchFn(`${client.baseUrl}/zones/${encodeURIComponent(zoneId)}/agents/${encodeURIComponent(agentSessionId)}`, {
    method: 'DELETE',
    headers: { authorization: `Bearer ${bearer}` },
  })
  if (!del.ok) {
    const text = await del.text()
    throw new Error(`coordinator DELETE /zones/${zoneId}/agents/${agentSessionId} failed: ${del.status} ${text}`)
  }
}

export async function createDelegation(client: CoordinatorClient, bearer: string, req: DelegationRequest): Promise<DelegationResponse> {
  const constraints = req.constraints
    ? {
        resources: req.constraints.resources,
        max_depth: req.constraints.maxDepth,
        max_hops: req.constraints.maxHops,
        ttl_seconds: req.constraints.ttlSeconds,
        budget: req.constraints.budget,
        policy_approved: req.constraints.policyApproved,
        expires_at: req.constraints.expiresAt,
        broad_reason: req.constraints.broadReason,
      }
    : undefined
  return call<DelegationResponse>(client, 'POST', `/zones/${encodeURIComponent(req.zoneId)}/delegations`, bearer, {
    issuer_application_id: req.issuerApplicationId,
    source_session_id: req.sourceSessionId,
    target_session_id: req.targetSessionId,
    receiver_application_id: req.receiverApplicationId,
    parent_edge_id: req.parentEdgeId,
    resource_id: req.resourceId ?? null,
    scopes: req.scopes,
    constraints,
    ttl_seconds: req.ttlSeconds,
  })
}

export async function heartbeatAgent(
  client: CoordinatorClient,
  bearer: string,
  zoneId: string,
  agentSessionId: string,
  status: 'starting' | 'healthy' | 'degraded' | 'unhealthy' = 'healthy',
): Promise<void> {
  await call<unknown>(
    client,
    'POST',
    `/zones/${encodeURIComponent(zoneId)}/agents/${encodeURIComponent(agentSessionId)}/heartbeat`,
    bearer,
    { status },
  )
}
